tools/src/main/scala/scala/scalanative/codegen/llvm/CodeGen.scala

```diff
@@ -68,52 +68,71 @@ object CodeGen {
       val env = assembly.map(defn => defn.name -> defn).toMap
       val workDir = VirtualDirectory.real(config.workDir)
 
+      def sourceFileOf(pos: Position): Position.SourceFile = {
+        if (pos == null || pos.isEmpty) new Position.SourceFile("__empty")
+        else pos.source
+      }
+
       // Partition into multiple LLVM IR files proportional to number
       // of available processesors. This prevents LLVM from optimizing
       // across IR module boundary unless LTO is turned on.
-      def separate(): Future[Seq[Path]] =
+      def separate(): Future[Seq[Path]] = {
         Future
-          .traverse(partitionBy(assembly, procs)(_.name.top).toSeq) {
+          .traverse(
+            partitionBy(assembly, procs)(x => sourceFileOf(x.pos)).toSeq
```

Comment:
It might be better to not seperate by exact path path of file, but rather by directory in which given file was defined. This way we should get the similar ammount of compilation units as when using packages to split defns. 
Keeping ammount of compilation units low, is important for getting optimal comilation times using LLVM as it was found out in https://github.com/scala-native/scala-native/pull/2777

Reply:
Thanks, now scala-native separates the .ll files per directory:

```
$ ls -l unit-tests/native/.3/target/scala-3.1.3/native-test/**/*.ll.o  | head -n 5 | awk '{print $9}'
unit-tests/native/.3/target/scala-3.1.3/native-test/Users/tanishiking/ghq/github.com/tanishiking/scala-native/auxlib/src/main/scala-3/scala/runtime/function.ll.o
unit-tests/native/.3/target/scala-3.1.3/native-test/Users/tanishiking/ghq/github.com/tanishiking/scala-native/auxlib/src/main/scala/scala/collection/concurrent.ll.o
unit-tests/native/.3/target/scala-3.1.3/native-test/Users/tanishiking/ghq/github.com/tanishiking/scala-native/auxlib/src/main/scala/scala/math.ll.o
unit-tests/native/.3/target/scala-3.1.3/native-test/Users/tanishiking/ghq/github.com/tanishiking/scala-native/auxlib/src/main/scala/scala/runtime.ll.o
unit-tests/native/.3/target/scala-3.1.3/native-test/Users/tanishiking/ghq/github.com/tanishiking/scala-native/auxlib/src/main/scala/scala/scalanative/runtime.ll.o
```

and this fixed wired issues in failing tests magically.

Currently, waiting for the CI for Windows.